Intermediate tour: Null safety chapter

sarahhaggarty opened the pull request:
This PR is to add the fourth chapter of the intermediate tour to the staging branch.

danil-pavlov commented on docs/topics/tour/kotlin-tour-intermediate-null-safety.md:
```diff
@@ -0,0 +1,658 @@
+specific type, this process is called **casting**. When a type is automatically cast, like when it's inferred, it's called
+**smart casting**.
+
+Before we explore how casting works, how can you check if an object has a certain type? For this, you can use the
```
Just a nitpick, I feel like the connection in the first sentence is not very clear, what about:

> Before we explore how casting works, let's learn how you can check if an object has a certain type.

or

> Before exploring how casting works, let's see how to check if an object has a certain type.

danil-pavlov commented on docs/topics/tour/kotlin-tour-intermediate-null-safety.md:
```diff
@@ -0,0 +1,658 @@
+* `is` checks if the object has the type and returns a boolean value.
+* `!is` checks if the object **doesn't** have the type and returns a boolean value.
+
+You've already seen this in practice when you learned about sealed classes in the [Open and other special classes](kotlin-tour-intermediate-classes-part-3.md) chapter.
```
This note breaks the flow a little bit; maybe move it under the code as a tip?

danil-pavlov commented on docs/topics/tour/kotlin-tour-intermediate-null-safety.md:
````diff
@@ -0,0 +1,658 @@
+}
+```
+
+In the first version, the example:
````
It might be even better to have these lists with explanations right next to the code
Plus, visually is would help to avoid to consequent lists

sarahhaggarty replied:
I restructured it as you suggest :)

danil-pavlov commented on docs/topics/tour/kotlin-tour-intermediate-null-safety.md:
```diff
@@ -0,0 +1,658 @@
+You are developing a notification system for an app where users can enable or disable different types of notifications.
+Complete the `getNotificationPreferences()` function so that:
+
+1. The `validUser` variable uses the `as?` operator to check if `user` is a `User` object. If it isn't, return an empty list.
```
```suggestion
1. The `validUser` variable uses the `as?` operator to check if `user` is an object of the `User` class. If it isn't, return an empty list.
```

danil-pavlov commented on docs/topics/tour/kotlin-tour-intermediate-null-safety.md:
```diff
@@ -0,0 +1,658 @@
+* [`maxOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/max-or-null.html), finds the highest value. If one doesn't exist, returns a `null` value.
+* [`minOrNull()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/min-or-null.html), finds the lowest value. If one doesn't exist, returns a `null` value.
+
+> The `singleOrNull()`, `maxOrNull()`, and `minOrNull()` functions are designed to be used with collections that **don't**
```
Again, I feel like the note here breaks the flow (and makes it look like the example is directly connected to the note, and not to the list), consider moving it lower

danil-pavlov reviewed (changes requested):
Great chapter! Just minor suggestions
Also,  I'm a bit concerned with the length of "Smart casts and safe casts" and "Early returns and the Elvis operator﻿". It's mostly because of code samples,  but it would be great to have subsections there (if it;s possible of course)

sarahhaggarty commented on docs/topics/tour/kotlin-tour-intermediate-null-safety.md:
```diff
@@ -0,0 +1,658 @@
+**smart casting**.
+
+Before we explore how casting works, how can you check if an object has a certain type? For this, you can use the
+`is` and `!is` operators with `when` or `if`:
```
In the beginner tour, we refer to `when` or `if` without calling them conditions. Can we do the same here? https://kotlinlang.org/docs/kotlin-tour-control-flow.html#conditional-expressions

danil-pavlov replied:
Maybe just expressions? It's just that in this sentence, `is` and `!is?` have a term, but `when` and `if` don't

sarahhaggarty replied:
I see now. I went with "conditional expressions". Since `when` can be an expression or a statement 😉

sarahhaggarty commented:
Thanks @danil-pavlov !
I added two new headers in "Smart casts and safe casts" but I couldn't find a way to add headers in "Early returns and the Elvis operator﻿". I also added more sample starts/ends to reduce the code sample sizes. Hopefully that makes the chapter seem a bit shorter. WDYT?
